### PR TITLE
ARROW-5853: [Python] Expose boolean filter kernel on Array

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
https://issues.apache.org/jira/browse/ARROW-5853

### CODE REVIEW

File: python/pyarrow/array.pxi [hunk 1]
```diff
@@ -824,6 +824,46 @@ cdef class Array(_PandasConvertible):
...
+          null,
+          "e"
+        ]
+        """
+        cdef:
+            cdef CDatum out
+
+        if mask.type.id != _Type_BOOL:
+            raise TypeError(
+                "The mask should be boolean type, got {}".format(mask.type))
```
Comment (jorisvandenbossche) [inline python/pyarrow/array.pxi]: I was wondering, is this a check that we might want to do on the C++ side? (you get a segfault otherwise)

Comment (pitrou) [reply python/pyarrow/array.pxi]: Yes, that would be preferable indeed.

Comment (jorisvandenbossche) [reply python/pyarrow/array.pxi]: OK, done

**Comment by jorisvandenbossche** [comment]
The travis failure is the java one (https://issues.apache.org/jira/browse/ARROW-6509)

File: python/pyarrow/tests/test_compute.py [hunk 2]
```diff
@@ -108,3 +111,29 @@ def test_take_dictionary(ordered):
...
+
+    mask = pa.array([True, False, False, True, None])
+    result = arr.filter(mask)
+    result.validate()
+
+    expected = pa.array([values[0], values[3], None], type=ty)
+    assert result.equals(expected)
+
+
+def test_filter_invalid():
```
Comment (bkietz) [inline python/pyarrow/tests/test_compute.py]: should this test be parameterized as well?

**Review by bkietz — CHANGES_REQUESTED** [review]
mostly LGTM

File: python/pyarrow/tests/test_compute.py (hunk 2 above)
Comment (jorisvandenbossche) [reply python/pyarrow/tests/test_compute.py]: Done

**Review by pitrou — APPROVED** [review]
LGTM

**Pull request closed without merging**